proposals/0166-dynamic-stack-frames.md

```diff
@@ -0,0 +1,173 @@
+problem, it would supply an unnecessary amount of memory to functions even 
+when they do not need them. In addition, such a solution would increase 
+pressure on the total memory available for the call stack. Either we would 
+need to increase the total allocation for the virtual machine or decrease the 
+maximum call depth.
+
+## New Terminology
+
+None.
+
+## Detailed Design
+
+Bringing dynamic stack frames to the Solana Bytecode Format and its 
+corresponding virtual machine entails changes in several aspects of the 
+execution environment.
+
+### SBF architecture modifications
+
+
+We will introduce a new register R11 in the virtual machine, which is going 
+to hold the stack pointer. The program must only write to such a register and 
+modify it through the `add64 reg, imm` (op code `0x07`) instruction. The 
+verifier must enforce these constraints on deployed programs. For further 
+information about the  changes in the ISA, refer to the 
+[SPF spec document](https://github.com/solana-labs/rbpf/blob/main/doc/bytecode.md).
+
+The R11 register must work in tandem with the R10 (frame pointer) register. 
+The former is write-only to the program, and the latter is read-only to the 
+program, forming a common design pattern in hardware engineering. More 
+details of this usage are in the following section.
+
+### Changes in the execution environment
+
+The R10 register must continue to hold the frame pointer, but we will manage 
+it differently. With fixed frames, when there is a function call we add 4096 
+to R10 and subtract it when the function returns. In the new scheme, we must 
+assign the value of R11 to R10 at function calls, and save R10’s former value 
+so that we can restore it when the function returns.
+
+The introduction of dynamic stack frames will change the direction of stack 
+growth. Presently, we stack frames on top of each other, but the memory usage 
+in them grows downward. In the new frame setting, both the placement of new 
+frames and the memory usage inside frames must be downward.
+
+The stack frame gaps feature, which creates a memory layout where frames are 
+interleaved with equally sized gaps, are not compatible with dynamic stack 
+frames and must be deactivated.
+
+### Changes in code generation
+
+In the compiler side, dynamic stack frames allow for some optimizations. 
+First, when a function does not need any stack allocated variable, code 
+generation must not create any instruction to modify R11. In addition, we 
+can stop using R5 as a stack spill register when a function call receives 
+more than five arguments. With dynamic stack frames, the compiler must use 
+registers R1 to R5 for the first five arguments and place remainder arguments 
+in the callee frame, instead of placing them in the caller’s frame. This new 
+call convention obviates the need to use R5 for retrieving the caller’s frame 
+pointer address to access those parameters.
+
+### Changes in the verifier
```

Comment:
I would simplify this section. ATM it seems to describe the whole process of validating registers, not just the changes due to this SIMD.

The verifier should include a new rule to accept R11 as a destination register under the following conditions:
1. Dynamic stack frames are enabled
2. The corresponding instruction is `add64_imm` (opcode `0x07`)
3. The corresponding value of `imm` is a multiple of the alignment (i.e. multiple of 64).

Reply:
As a comment to my comment :) Based on the current code, I'd simply add an extra "and" to the "if" condition that returns Ok() for R11. Therefore I don't think we need an error for alignment. Either R11 is allowed (if all 3 conditions pass), or it's not.

Reply:
I re-wrote the section, based on the new design we devised with only R10.